# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [61]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. 

In [62]:
austin = pd.read_csv('/Users/GiadaSartori/Documents/IRONHACK/Lab/Week7/lab-intro-to-ml/your-code/austin_weather.csv')
austin

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0     2013-12-21         74        60        45            67           49   
1     2013-12-22         56        48        39            43           36   
2     2013-12-23         58        45        32            31           27   
3     2013-12-24         61        46        31            36           28   
4     2013-12-25         58        50        41            44           40   
...          ...        ...       ...       ...           ...          ...   
1314  2017-07-27        103        89        75            71           67   
1315  2017-07-28        105        91        76            71           64   
1316  2017-07-29        107        92        77            72           64   
1317  2017-07-30        106        93        79            70           68   
1318  2017-07-31         99        88        77            66           61   

     DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0              43                  93                 75                 57   
1              28                  93                 68                 43   
2              23                  76                 52                 27   
3              21                  89                 56                 22   
4              36                  86                 71                 56   
...           ...                 ...                ...                ...   
1314           61                  82                 54                 25   
1315           55                  87                 54                 20   
1316           55                  82                 51                 19   
1317           63                  69                 48                 27   
1318           54                  64                 43                 22   

      ... SeaLevelPressureAvgInches SeaLevelPressureLowInches  \
0     ...                     29.68                     29.59   
1     ...                     30.13                     29.87   
2     ...                     30.49                     30.41   
3     ...                     30.45                      30.3   
4     ...                     30.33                     30.27   
...   ...                       ...                       ...   
1314  ...                     29.97                     29.88   
1315  ...                      29.9                     29.81   
1316  ...                     29.86                     29.79   
1317  ...                     29.91                     29.87   
1318  ...                     29.97                     29.91   

     VisibilityHighMiles VisibilityAvgMiles VisibilityLowMiles WindHighMPH  \
0                     10                  7                  2          20   
1                     10                 10                  5          16   
2                     10                 10                 10           8   
3                     10                 10                  7          12   
4                     10                 10                  7          10   
...                  ...                ...                ...         ...   
1314                  10                 10                 10          12   
1315                  10                 10                 10          14   
1316                  10                 10                 10          12   
1317                  10                 10                 10          13   
1318                  10                 10                 10          12   

     WindAvgMPH WindGustMPH PrecipitationSumInches               Events  
0             4          31                   0.46  Rain , Thunderstorm  
1             6          25                      0                       
2             3          12                      0                       
3             4          20                      0                       
4             2          16           

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [63]:
# Examine the dataset

# 1. Which types of data our austin dataframe contains

austin.dtypes


Date                          object
TempHighF                      int64
TempAvgF                       int64
TempLowF                       int64
DewPointHighF                 object
DewPointAvgF                  object
DewPointLowF                  object
HumidityHighPercent           object
HumidityAvgPercent            object
HumidityLowPercent            object
SeaLevelPressureHighInches    object
SeaLevelPressureAvgInches     object
SeaLevelPressureLowInches     object
VisibilityHighMiles           object
VisibilityAvgMiles            object
VisibilityLowMiles            object
WindHighMPH                   object
WindAvgMPH                    object
WindGustMPH                   object
PrecipitationSumInches        object
Events                        object
dtype: object

In [64]:
#Size of our dataframe austin

austin.shape

(1319, 21)

In [65]:
# Verify if there are any null values in the dataframe austin
austin.isnull().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
PrecipitationSumInches        0
Events                        0
dtype: int64

In [66]:
austin.head(5)

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

In [67]:
# Check the unique categorical values of the column Events
austin['Events'].unique()

array(['Rain , Thunderstorm', ' ', 'Rain', 'Fog', 'Rain , Snow',
       'Fog , Rain', 'Thunderstorm', 'Fog , Rain , Thunderstorm',
       'Fog , Thunderstorm'], dtype=object)

In [68]:
# Check the maximum value of the TempHighF
max_temp = austin['TempHighF'].max()
max_temp

107

In [69]:
# Check the index of the max value of TempHighF
austin[austin['TempHighF'] == max_temp]

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
1316  2017-07-29        107        92        77            72           64   

     DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
1316           55                  82                 51                 19   

      ... SeaLevelPressureAvgInches SeaLevelPressureLowInches  \
1316  ...                     29.86                     29.79   

     VisibilityHighMiles VisibilityAvgMiles VisibilityLowMiles WindHighMPH  \
1316                  10                 10                 10          12   

     WindAvgMPH WindGustMPH PrecipitationSumInches Events  
1316          4          17                      0         

[1 rows x 21 columns]

In [70]:
# Check the minimum value of the TempLowF
min_temp = austin['TempLowF'].min()
min_temp

19

In [71]:
# Check the index of the min value of TempLowF
austin[austin['TempLowF'] == min_temp]

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
1113  2017-01-07         44        32        19            13           11   

     DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
1113            6                  73                 48                 22   

      ... SeaLevelPressureAvgInches SeaLevelPressureLowInches  \
1113  ...                     30.74                     30.61   

     VisibilityHighMiles VisibilityAvgMiles VisibilityLowMiles WindHighMPH  \
1113                  10                 10                 10          12   

     WindAvgMPH WindGustMPH PrecipitationSumInches Events  
1113          4          21                      0         

[1 rows x 21 columns]

In [72]:
# Check the average temperature of the TempAvgF from the dates where the maximum and minimum temperatures were recorded

austin[austin["Date"].between('2017-01-07','2017-07-29')]["TempAvgF"].mean()

73.74509803921569

**Findings from Giada:** 

1. We have 21 columns in the dataset austin, of which:
    - TempHighF, TempAvgF, and TempLowF have numerical values, which are interger
    - All columns, except the above mentioned, contain objects
    - Events contain 10 unique categorical values

2. Numerical columns have different units, which can have an impact on the accuracy of the score of predicted model

3. Date column can be split into potential 3 different columns, year, month and date

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [73]:
austin['DewPointAvgF'].unique()

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

**Observations:** I thinks pandas is treating the column 'DewPointAvgF' as an object, instead of an interger, because one of the unique value is a '-'.

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

**Function:** pd.to_numeric(s, errors='coerce')

In [74]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

1. Apply numeric values to the columns with the wrong types

In [75]:
austin_2 = austin[wrong_type_columns].apply(pd.to_numeric, errors='coerce')
austin_2

DewPointHighF  DewPointAvgF  DewPointLowF  HumidityHighPercent  \
0              67.0          49.0          43.0                 93.0   
1              43.0          36.0          28.0                 93.0   
2              31.0          27.0          23.0                 76.0   
3              36.0          28.0          21.0                 89.0   
4              44.0          40.0          36.0                 86.0   
...             ...           ...           ...                  ...   
1314           71.0          67.0          61.0                 82.0   
1315           71.0          64.0          55.0                 87.0   
1316           72.0          64.0          55.0                 82.0   
1317           70.0          68.0          63.0                 69.0   
1318           66.0          61.0          54.0                 64.0   

      HumidityAvgPercent  HumidityLowPercent  SeaLevelPressureHighInches  \
0                   75.0                57.0                       29.86   
1                   68.0                43.0                       30.41   
2                   52.0                27.0                       30.56   
3                   56.0                22.0                       30.56   
4                   71.0                56.0                       30.41   
...                  ...                 ...                         ...   
1314                54.0                25.0                       30.04   
1315                54.0                20.0                       29.97   
1316                51.0                19.0                       29.91   
1317                48.0                27.0                       29.96   
1318                43.0                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH  PrecipitationSumInches  
0            20.0         4.0         31.0                    0.46  
1            16.0         6.0         25.0                    0.00  
2             8.0         3.0         12.0                    0.00  
3            12.0         4.0         20.0                    0.00  
4            10.0         2.0         16.0                     NaN  
...           ...         ...          ...                     ...  
1314         12.0         5.0         21.0                    0.00  
1315         14.0         5.0         20.0                    0.00  
1316         12.0         4.0    

2. Drop columns which types are object, from original dataset which is austin --> df.drop(columns=['B', 'C'])

In [76]:
austin = austin.drop(columns=['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches'])

austin

Date  TempHighF  TempAvgF  TempLowF               Events
0     2013-12-21         74        60        45  Rain , Thunderstorm
1     2013-12-22         56        48        39                     
2     2013-12-23         58        45        32                     
3     2013-12-24         61        46        31                     
4     2013-12-25         58        50        41                     
...          ...        ...       ...       ...                  ...
1314  2017-07-27        103        89        75                     
1315  2017-07-28        105        91        76                     
1316  2017-07-29        107        92        77                     
1317  2017-07-30        106        93        79                     
1318  2017-07-31         99        88        77                     

[1319 rows x 5 columns]

In [77]:
austin_transformed = pd.merge(left=austin, right=austin_2, left_index=True, right_index=True)
austin_transformed

Date  TempHighF  TempAvgF  TempLowF               Events  \
0     2013-12-21         74        60        45  Rain , Thunderstorm   
1     2013-12-22         56        48        39                        
2     2013-12-23         58        45        32                        
3     2013-12-24         61        46        31                        
4     2013-12-25         58        50        41                        
...          ...        ...       ...       ...                  ...   
1314  2017-07-27        103        89        75                        
1315  2017-07-28        105        91        76                        
1316  2017-07-29        107        92        77                        
1317  2017-07-30        106        93        79                        
1318  2017-07-31         99        88        77                        

      DewPointHighF  DewPointAvgF  DewPointLowF  HumidityHighPercent  \
0              67.0          49.0          43.0                 93.0   
1              43.0          36.0          28.0                 93.0   
2              31.0          27.0          23.0                 76.0   
3              36.0          28.0          21.0                 89.0   
4              44.0          40.0          36.0                 86.0   
...             ...           ...           ...                  ...   
1314           71.0          67.0          61.0                 82.0   
1315           71.0          64.0          55.0                 87.0   
1316           72.0          64.0          55.0                 82.0   
1317           70.0          68.0          63.0                 69.0   
1318           66.0          61.0          54.0                 64.0   

      HumidityAvgPercent  ...  SeaLevelPressureHighInches  \
0                   75.0  ...                       29.86   
1                   68.0  ...                       30.41   
2                   52.0  ...                       30.56   
3                   56.0  ...                       30.56   
4                   71.0  ...                       30.41   
...                  ...  ...                         ...   
1314                54.0  ...                       30.04   
1315                54.0  ...                       29.97   
1316                51.0  ...                       29.91   
1317                48.0  ...                       29.96   
1318                43.0  ...                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0

In [78]:
austin_transformed.dtypes

Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
Events                         object
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
dtype: object

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [79]:
# See the sum of missing values in our dataset austin_transformed

austin_transformed.isnull().sum()


Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
Events                          0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
dtype: int64

In [80]:
# Find the rows of the missing data of our dataset austin_transformed

missing_values = austin_transformed.isnull().any(axis=1)
missing_values

0       False
1       False
2       False
3       False
4        True
        ...  
1314    False
1315    False
1316    False
1317    False
1318    False
Length: 1319, dtype: bool

In [81]:
# We localize the rows in dataset austin_transformed where rows contain missin values

missing_values_df = austin_transformed.loc[missing_values] #or == True
missing_values_df


Date  TempHighF  TempAvgF  TempLowF               Events  \
4     2013-12-25         58        50        41                        
6     2013-12-27         60        53        45                        
7     2013-12-28         62        51        40                        
42    2014-02-01         76        66        55                 Rain   
51    2014-02-10         60        48        35                 Rain   
...          ...        ...       ...       ...                  ...   
1269  2017-06-12         94        85        75                 Rain   
1288  2017-07-01         98        89        79                        
1304  2017-07-17         98        88        77  Rain , Thunderstorm   
1308  2017-07-21        104        91        77                        
1313  2017-07-26        103        91        78                        

      DewPointHighF  DewPointAvgF  DewPointLowF  HumidityHighPercent  \
4              44.0          40.0          36.0                 86.0   
6              41.0          39.0          37.0                 83.0   
7              43.0          39.0          33.0                 92.0   
42             62.0          59.0          41.0                 81.0   
51             49.0          36.0          30.0                 82.0   
...             ...           ...           ...                  ...   
1269           72.0          68.0          62.0                 87.0   
1288           76.0          72.0          66.0                 85.0   
1304           75.0          71.0          66.0                 88.0   
1308           74.0          68.0          61.0                 85.0   
1313           75.0          70.0          61.0                 91.0   

      HumidityAvgPercent  ...  SeaLevelPressureHighInches  \
4                   71.0  ...                       30.41   
6                   65.0  ...                       30.46   
7                   64.0  ...                       30.33   
42                  71.0  ...                       29.91   
51                  74.0  ...                       30.23   
...                  ...  ...                         ...   
1269                62.0  ...                       29.96   
1288                62.0  ...                       30.13   
1304                63.0  ...                       29.94   
1308                56.0  ...                       30.01   
1313                58.0  ...                       30.13   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
4                         30.33                      30.27   
6                         30.39                      30.34   
7                         30.17                      30.04   
42                        29.81                      29.75   
51                        30.15                      30.02   
...                         ...                        ...   
1269                      29.89                      29.81   
1288                      30.02                      29.92   
1304                      29.90                      29.84   
1308                      29.95                      29.86   
1313                      30.03                      29.93   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
4                    10.0                10.0                 7.0   
6                    10.0                 9.0                 7.0   
7                    10.0                10.0                 7.0   
42                   10.0                10.0                 9.0   
51                   10.0                 8.0                 4.0   
...                   ...                 ...                 ...   
1269                 10.0                10.0                 2.0   
1288                 10.0                 9.0                 6.0   
1304                 10.0                10.0                10.0   
1308                 10.0                10.0                10.0   
1313                 10.0                10.0

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [82]:
austin_transformed.shape

(1319, 21)

In [83]:
missing_values_df.shape

(136, 21)

#### Calculate the ratio of missing rows to total rows

In [84]:
ratio = (136/1319)*100
ratio


10.310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [85]:
precipation_ratio = (124/1319)*100
precipation_ratio

9.401061410159212

Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [86]:
austin_transformed.head(3)

Date  TempHighF  TempAvgF  TempLowF               Events  \
0  2013-12-21         74        60        45  Rain , Thunderstorm   
1  2013-12-22         56        48        39                        
2  2013-12-23         58        45        32                        

   DewPointHighF  DewPointAvgF  DewPointLowF  HumidityHighPercent  \
0           67.0          49.0          43.0                 93.0   
1           43.0          36.0          28.0                 93.0   
2           31.0          27.0          23.0                 76.0   

   HumidityAvgPercent  ...  SeaLevelPressureHighInches  \
0                75.0  ...                       29.86   
1                68.0  ...                       30.41   
2                52.0  ...                       30.56   

   SeaLevelPressureAvgInches  SeaLevelPressureLowInches  VisibilityHighMiles  \
0                      29.68                      29.59                 10.0   
1                      30.13                      29.87                 10.0   
2                      30.49                      30.41                 10.0   

   VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0                 7.0                 2.0         20.0         4.0   
1                10.0                 5.0         16.0         6.0   
2                10.0                10.0          8.0         3.0   

   WindGustMPH  PrecipitationSumInches  
0         31.0                    0.46  
1         25.0                    0.00  
2         12.0                    0.00  

[3 rows x 21 columns]

In [87]:
austin_transformed = austin_transformed.drop(columns=['PrecipitationSumInches'])

In [88]:
# Print `austin_transformed` to confirm the column is indeed removed

austin_transformed

Date  TempHighF  TempAvgF  TempLowF               Events  \
0     2013-12-21         74        60        45  Rain , Thunderstorm   
1     2013-12-22         56        48        39                        
2     2013-12-23         58        45        32                        
3     2013-12-24         61        46        31                        
4     2013-12-25         58        50        41                        
...          ...        ...       ...       ...                  ...   
1314  2017-07-27        103        89        75                        
1315  2017-07-28        105        91        76                        
1316  2017-07-29        107        92        77                        
1317  2017-07-30        106        93        79                        
1318  2017-07-31         99        88        77                        

      DewPointHighF  DewPointAvgF  DewPointLowF  HumidityHighPercent  \
0              67.0          49.0          43.0                 93.0   
1              43.0          36.0          28.0                 93.0   
2              31.0          27.0          23.0                 76.0   
3              36.0          28.0          21.0                 89.0   
4              44.0          40.0          36.0                 86.0   
...             ...           ...           ...                  ...   
1314           71.0          67.0          61.0                 82.0   
1315           71.0          64.0          55.0                 87.0   
1316           72.0          64.0          55.0                 82.0   
1317           70.0          68.0          63.0                 69.0   
1318           66.0          61.0          54.0                 64.0   

      HumidityAvgPercent  HumidityLowPercent  SeaLevelPressureHighInches  \
0                   75.0                57.0                       29.86   
1                   68.0                43.0                       30.41   
2                   52.0                27.0                       30.56   
3                   56.0                22.0                       30.56   
4                   71.0                56.0                       30.41   
...                  ...                 ...                         ...   
1314                54.0                25.0                       30.04   
1315                54.0                20.0                       29.97   
1316                51.0                19.0                       29.91   
1317                48.0                27.0                       29.96   
1318                43.0                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
131

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [89]:
austin_transformed_fixed = austin_transformed.interpolate()
austin_transformed_fixed

Date  TempHighF  TempAvgF  TempLowF               Events  \
0     2013-12-21         74        60        45  Rain , Thunderstorm   
1     2013-12-22         56        48        39                        
2     2013-12-23         58        45        32                        
3     2013-12-24         61        46        31                        
4     2013-12-25         58        50        41                        
...          ...        ...       ...       ...                  ...   
1314  2017-07-27        103        89        75                        
1315  2017-07-28        105        91        76                        
1316  2017-07-29        107        92        77                        
1317  2017-07-30        106        93        79                        
1318  2017-07-31         99        88        77                        

      DewPointHighF  DewPointAvgF  DewPointLowF  HumidityHighPercent  \
0              67.0          49.0          43.0                 93.0   
1              43.0          36.0          28.0                 93.0   
2              31.0          27.0          23.0                 76.0   
3              36.0          28.0          21.0                 89.0   
4              44.0          40.0          36.0                 86.0   
...             ...           ...           ...                  ...   
1314           71.0          67.0          61.0                 82.0   
1315           71.0          64.0          55.0                 87.0   
1316           72.0          64.0          55.0                 82.0   
1317           70.0          68.0          63.0                 69.0   
1318           66.0          61.0          54.0                 64.0   

      HumidityAvgPercent  HumidityLowPercent  SeaLevelPressureHighInches  \
0                   75.0                57.0                       29.86   
1                   68.0                43.0                       30.41   
2                   52.0                27.0                       30.56   
3                   56.0                22.0                       30.56   
4                   71.0                56.0                       30.41   
...                  ...                 ...                         ...   
1314                54.0                25.0                       30.04   
1315                54.0                20.0                       29.97   
1316                51.0                19.0                       29.91   
1317                48.0                27.0                       29.96   
1318                43.0                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
131

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [90]:
austin_transformed_fixed.isna().sum()


Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
Events                        0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
dtype: int64

In [91]:
austin_transformed.isna().sum()

Date                           0
TempHighF                      0
TempAvgF                       0
TempLowF                       0
Events                         0
DewPointHighF                  7
DewPointAvgF                   7
DewPointLowF                   7
HumidityHighPercent            2
HumidityAvgPercent             2
HumidityLowPercent             2
SeaLevelPressureHighInches     3
SeaLevelPressureAvgInches      3
SeaLevelPressureLowInches      3
VisibilityHighMiles           12
VisibilityAvgMiles            12
VisibilityLowMiles            12
WindHighMPH                    2
WindAvgMPH                     2
WindGustMPH                    4
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [92]:
austin_transformed_fixed['Events'].value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: Events, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

We clearly see that the largest amount of events that had happended in a single day are three, which are fog, rain and thunderstorm.
We now want to know in which days, which are 33, those events had happended. 

In [93]:
austin_transformed_fixed_3events = austin_transformed_fixed.loc[austin_transformed_fixed['Events'] == 'Fog , Rain , Thunderstorm']
austin_transformed_fixed_3events

Date  TempHighF  TempAvgF  TempLowF                     Events  \
114   2014-04-14         73        58        43  Fog , Rain , Thunderstorm   
142   2014-05-12         82        72        62  Fog , Rain , Thunderstorm   
157   2014-05-27         73        68        63  Fog , Rain , Thunderstorm   
185   2014-06-24         93        84        74  Fog , Rain , Thunderstorm   
186   2014-06-25         89        80        71  Fog , Rain , Thunderstorm   
208   2014-07-17         89        82        74  Fog , Rain , Thunderstorm   
257   2014-09-04         99        86        73  Fog , Rain , Thunderstorm   
271   2014-09-18         90        81        72  Fog , Rain , Thunderstorm   
336   2014-11-22         67        63        58  Fog , Rain , Thunderstorm   
438   2015-03-04         70        52        34  Fog , Rain , Thunderstorm   
483   2015-04-18         77        66        54  Fog , Rain , Thunderstorm   
500   2015-05-05         83        75        66  Fog , Rain , Thunderstorm   
508   2015-05-13         70        67        64  Fog , Rain , Thunderstorm   
518   2015-05-23         87        77        66  Fog , Rain , Thunderstorm   
520   2015-05-25         82        74        66  Fog , Rain , Thunderstorm   
540   2015-06-14         86        79        72  Fog , Rain , Thunderstorm   
542   2015-06-16         83        78        73  Fog , Rain , Thunderstorm   
543   2015-06-17         89        81        72  Fog , Rain , Thunderstorm   
545   2015-06-19         84        78        71  Fog , Rain , Thunderstorm   
546   2015-06-20         83        78        72  Fog , Rain , Thunderstorm   
547   2015-06-21         86        79        72  Fog , Rain , Thunderstorm   
556   2015-06-30         94        84        73  Fog , Rain , Thunderstorm   
628   2015-09-10         96        84        72  Fog , Rain , Thunderstorm   
678   2015-10-30         75        71        67  Fog , Rain , Thunderstorm   
848   2016-04-17         78        71        64  Fog , Rain , Thunderstorm   
861   2016-04-30         86        76        66  Fog , Rain , Thunderstorm   
889   2016-05-28         91        78        65  Fog , Rain , Thunderstorm   
949   2016-07-27         92        83        74  Fog , Rain , Thunderstorm   
969   2016-08-16         83        77        71  Fog , Rain , Thunderstorm   
1009  2016-09-25         88        79        70  Fog , Rain , Thunderstorm   
1048  2016-11-03         80        74        68  Fog , Rain , Thunderstorm   
1123  2017-01-17         59        54        48  Fog , Rain , Thunderstorm   
1262  2017-06-05         90        78        66  Fog , Rain , Thunderstorm   

      DewPointHighF  DewPointAvgF  DewPointLowF  HumidityHighPercent  \
114            69.0          46.0          32.0                 90.0   
142            76.0          70.0          61.0                 94.0   
157            67.0          64.0          61.0                 93.0   
185            74.0          72.0          69.0                 94.0   
186            76.0          72.0          68.0                 94.0   
208            76.0          74.0          70.0                 94.0   
257            77.0          74.0          67.0                 91.0   
271            75.0          73.0          70.0                100.0   
336            64.0          62.0          56.0                100.0   
438            67.0          56.0          32.0                100.0   
483            66.0          63.0          53.0                100.0   
500            70.0          67.0          62.0                100.0   
508            70.0          66.0          64.0                100.0   
518            74.0          71.0          66.0                100.0   
520            74.0          69.0          66.0                100.0   
540            76.0          73.0          71.0                100.0   
542            75.0          73.0          70.0                 94.0   
543            77.0          74.0          71.0                

In [94]:
austin_transformed_fixed_3events.shape

(33, 20)

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [95]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

austin_transformed_fixed['Snow'] = 0 
austin_transformed_fixed['Fog'] = 0
austin_transformed_fixed['Rain'] = 0
austin_transformed_fixed['Thunderstorm'] = 0


# Print your new dataframe to check whether new columns have been created:

austin_transformed_fixed

Date  TempHighF  TempAvgF  TempLowF               Events  \
0     2013-12-21         74        60        45  Rain , Thunderstorm   
1     2013-12-22         56        48        39                        
2     2013-12-23         58        45        32                        
3     2013-12-24         61        46        31                        
4     2013-12-25         58        50        41                        
...          ...        ...       ...       ...                  ...   
1314  2017-07-27        103        89        75                        
1315  2017-07-28        105        91        76                        
1316  2017-07-29        107        92        77                        
1317  2017-07-30        106        93        79                        
1318  2017-07-31         99        88        77                        

      DewPointHighF  DewPointAvgF  DewPointLowF  HumidityHighPercent  \
0              67.0          49.0          43.0                 93.0   
1              43.0          36.0          28.0                 93.0   
2              31.0          27.0          23.0                 76.0   
3              36.0          28.0          21.0                 89.0   
4              44.0          40.0          36.0                 86.0   
...             ...           ...           ...                  ...   
1314           71.0          67.0          61.0                 82.0   
1315           71.0          64.0          55.0                 87.0   
1316           72.0          64.0          55.0                 82.0   
1317           70.0          68.0          63.0                 69.0   
1318           66.0          61.0          54.0                 64.0   

      HumidityAvgPercent  ...  VisibilityHighMiles  VisibilityAvgMiles  \
0                   75.0  ...                 10.0                 7.0   
1                   68.0  ...                 10.0                10.0   
2                   52.0  ...                 10.0                10.0   
3                   56.0  ...                 10.0                10.0   
4                   71.0  ...                 10.0                10.0   
...                  ...  ...                  ...                 ...   
1314                54.0  ...                 10.0                10.0   
1315                54.0  ...                 10.0                10.0   
1316                51.0  ...                 10.0                10.0   
1317                48.0  ...                 10.0                10.0   
1318                43.0  ...                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
0                    2.0         20.0         4.0         31.0     0    0   
1                    5.0         16.0         6.0         25.0     0    0   
2                   10.0          8.0         3.0         12.0     0    0   
3                    7.0         12.0         4.0         20.0     0    0   
4                    7.0         10.0         2.0         16.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
1314                10.0         12.0         5.0         21.0     0    0   
1315                10.0         14.0         5.0         20.0     0    0   
1316                10.0         12.0         4.0         17.0     0    0   
1317                10.0         13.0         4.0         20.0     0    0   
1318                10.0         12.0         4.0         20.0     0    0   

      Rain  Thunderstorm  
0        0             0  
1        0             0  
2        0             0  
3        0             0  
4        0             0  
...    ...           ...  
1314     0             0  
1315     0             0  
1316     0             0  
1317     0             0  
1318     0             0  

[1319 rows x 24 columns]

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [96]:
austin_transformed_fixed['Rain'] = austin_transformed_fixed['Events'].str.contains('Rain').astype(int)
austin_transformed_fixed['Rain']


0       1
1       0
2       0
3       0
4       0
       ..
1314    0
1315    0
1316    0
1317    0
1318    0
Name: Rain, Length: 1319, dtype: int64

In [97]:
austin_transformed_fixed['Snow'] = austin_transformed_fixed['Events'].str.contains('Snow').astype(int)
austin_transformed_fixed['Snow']

0       0
1       0
2       0
3       0
4       0
       ..
1314    0
1315    0
1316    0
1317    0
1318    0
Name: Snow, Length: 1319, dtype: int64

In [98]:
austin_transformed_fixed['Fog'] = austin_transformed_fixed['Events'].str.contains('Fog').astype(int)
austin_transformed_fixed['Fog']

0       0
1       0
2       0
3       0
4       0
       ..
1314    0
1315    0
1316    0
1317    0
1318    0
Name: Fog, Length: 1319, dtype: int64

In [99]:
austin_transformed_fixed['Thunderstorm'] = austin_transformed_fixed['Events'].str.contains('Thunderstorm').astype(int)
austin_transformed_fixed['Thunderstorm']

0       1
1       0
2       0
3       0
4       0
       ..
1314    0
1315    0
1316    0
1317    0
1318    0
Name: Thunderstorm, Length: 1319, dtype: int64

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [100]:
austin_transformed_fixed

Date  TempHighF  TempAvgF  TempLowF               Events  \
0     2013-12-21         74        60        45  Rain , Thunderstorm   
1     2013-12-22         56        48        39                        
2     2013-12-23         58        45        32                        
3     2013-12-24         61        46        31                        
4     2013-12-25         58        50        41                        
...          ...        ...       ...       ...                  ...   
1314  2017-07-27        103        89        75                        
1315  2017-07-28        105        91        76                        
1316  2017-07-29        107        92        77                        
1317  2017-07-30        106        93        79                        
1318  2017-07-31         99        88        77                        

      DewPointHighF  DewPointAvgF  DewPointLowF  HumidityHighPercent  \
0              67.0          49.0          43.0                 93.0   
1              43.0          36.0          28.0                 93.0   
2              31.0          27.0          23.0                 76.0   
3              36.0          28.0          21.0                 89.0   
4              44.0          40.0          36.0                 86.0   
...             ...           ...           ...                  ...   
1314           71.0          67.0          61.0                 82.0   
1315           71.0          64.0          55.0                 87.0   
1316           72.0          64.0          55.0                 82.0   
1317           70.0          68.0          63.0                 69.0   
1318           66.0          61.0          54.0                 64.0   

      HumidityAvgPercent  ...  VisibilityHighMiles  VisibilityAvgMiles  \
0                   75.0  ...                 10.0                 7.0   
1                   68.0  ...                 10.0                10.0   
2                   52.0  ...                 10.0                10.0   
3                   56.0  ...                 10.0                10.0   
4                   71.0  ...                 10.0                10.0   
...                  ...  ...                  ...                 ...   
1314                54.0  ...                 10.0                10.0   
1315                54.0  ...                 10.0                10.0   
1316                51.0  ...                 10.0                10.0   
1317                48.0  ...                 10.0                10.0   
1318                43.0  ...                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
0                    2.0         20.0         4.0         31.0     0    0   
1                    5.0         16.0         6.0         25.0     0    0   
2                   10.0          8.0         3.0         12.0     0    0   
3                    7.0         12.0         4.0         20.0     0    0   
4                    7.0         10.0         2.0         16.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
1314                10.0         12.0         5.0         21.0     0    0   
1315                10.0         14.0         5.0         20.0     0    0   
1316                10.0         12.0         4.0         17.0     0    0   
1317                10.0         13.0         4.0         20.0     0    0   
1318                10.0         12.0         4.0         20.0     0    0   

      Rain  Thunderstorm  
0        1             1  
1        0             0  
2        0             0  
3        0             0  
4        0             0  
...    ...           ...  
1314     0             0  
1315     0             0  
1316     0             0  
1317     0             0  
1318     0             0  

[1319 rows x 24 columns]

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [101]:
austin_transformed_fixed = austin_transformed_fixed.drop(columns=['Events'])
austin_transformed_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityHighMiles  VisibilityAvgMiles  \
0                   57.0  ...                 10.0                 7.0   
1                   43.0  ...                 10.0                10.0   
2                   27.0  ...                 10.0                10.0   
3                   22.0  ...                 10.0                10.0   
4                   56.0  ...                 10.0                10.0   
...                  ...  ...                  ...                 ...   
1314                25.0  ...                 10.0                10.0   
1315                20.0  ...                 10.0                10.0   
1316                19.0  ...                 10.0                10.0   
1317                27.0  ...                 10.0                10.0   
1318                22.0  ...                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
0                    2.0         20.0         4.0         31.0     0    0   
1                    5.0         16.0         6.0         25.0     0    0   
2                   10.0          8.0         3.0         12.0     0    0   
3                    7.0         12.0         4.0         20.0     0    0   
4                    7.0         10.0         2.0         16.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
1314                10.0         12.0         5.0         21.0     0    0   
1315                10.0         14.0         5.0         20.0     0    0   
1316                10.0         12.0         4.0         17.0     0    0   
1317                10.0         13.0         4.0         20.0     0    0   
1318                10.0         12.0         4.0         20.0     0    0   

      Rain  Thunderstorm  
0        1             1  
1        0             0  
2        0             0  
3        0             0  
4        0             0  
...    ...           ...  
1314     0             0  
1315     0             0  
1316     0             0  
1317     0             0  
1318     0             0  

[1319 rows x 23 columns]

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [102]:
from datetime import date
from datetime import datetime

def convert_date_to_ordinal(date):
    return datetime.strptime(date, '%Y-%M-%d').toordinal()


In [103]:
austin_transformed_fixed['Date'] = austin_transformed_fixed['Date'].apply(convert_date_to_ordinal)

# We apply the function to the column of the dataframe in scope, in order to transform the entire each single row of column 'Date' 

#### Print `austin_fixed` to check your `Date` column.

In [104]:
austin_transformed_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  734889         74        60        45           67.0          49.0   
1  734890         56        48        39           43.0          36.0   
2  734891         58        45        32           31.0          27.0   
3  734892         61        46        31           36.0          28.0   
4  734893         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0  ...                 10.0                 7.0                 2.0   
1  ...                 10.0                10.0                 5.0   
2  ...                 10.0                10.0                10.0   
3  ...                 10.0                10.0                 7.0   
4  ...                 10.0                10.0                 7.0   

   WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0         20.0         4.0         31.0     0    0     1             1  
1         16.0         6.0         25.0     0    0     0             0  
2          8.0         3.0         12.0     0    0     0             0  
3         12.0         4.0         20.0     0    0     0             0  
4         10.0         2.0         16.0     0    0     0             0  

[5 rows x 23 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [111]:
austin_transformed_fixed.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  734889         74        60        45           67.0          49.0   
1  734890         56        48        39           43.0          36.0   
2  734891         58        45        32           31.0          27.0   
3  734892         61        46        31           36.0          28.0   
4  734893         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0  ...                 10.0                 7.0                 2.0   
1  ...                 10.0                10.0                 5.0   
2  ...                 10.0                10.0                10.0   
3  ...                 10.0                10.0                 7.0   
4  ...                 10.0                10.0                 7.0   

   WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0         20.0         4.0         31.0     0    0     1             1  
1         16.0         6.0         25.0     0    0     0             0  
2          8.0         3.0         12.0     0    0     0             0  
3         12.0         4.0         20.0     0    0     0             0  
4         10.0         2.0         16.0     0    0     0             0  

[5 rows x 23 columns]

In [137]:
len(austin_transformed_fixed)

1319

In [120]:
# Y assignation

y = austin_transformed_fixed[['TempAvgF']]
y

TempAvgF
0           60
1           48
2           45
3           46
4           50
...        ...
1314        89
1315        91
1316        92
1317        93
1318        88

[1319 rows x 1 columns]

In [121]:
# X assignation

X = austin_transformed_fixed[['Date', 'TempHighF', 'TempLowF', 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']]

X

Date  TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
0     734889         74        45           67.0          49.0          43.0   
1     734890         56        39           43.0          36.0          28.0   
2     734891         58        32           31.0          27.0          23.0   
3     734892         61        31           36.0          28.0          21.0   
4     734893         58        41           44.0          40.0          36.0   
...      ...        ...       ...            ...           ...           ...   
1314  736356        103        75           71.0          67.0          61.0   
1315  736357        105        76           71.0          64.0          55.0   
1316  736358        107        77           72.0          64.0          55.0   
1317  736359        106        79           70.0          68.0          63.0   
1318  736360         99        77           66.0          61.0          54.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0                    93.0                75.0                57.0   
1                    93.0                68.0                43.0   
2                    76.0                52.0                27.0   
3                    89.0                56.0                22.0   
4                    86.0                71.0                56.0   
...                   ...                 ...                 ...   
1314                 82.0                54.0                25.0   
1315                 87.0                54.0                20.0   
1316                 82.0                51.0                19.0   
1317                 69.0                48.0                27.0   
1318                 64.0                43.0                22.0   

      SeaLevelPressureHighInches  ...  VisibilityHighMiles  \
0                          29.86  ...                 10.0   
1                          30.41  ...                 10.0   
2                          30.56  ...                 10.0   
3                          30.56  ...                 10.0   
4                          30.41  ...                 10.0   
...                          ...  ...                  ...   
1314                       30.04  ...                 10.0   
1315                       29.97  ...                 10.0   
1316                       29.91  ...                 10.0   
1317                       29.96  ...                 10.0   
1318                       30.04  ...                 10.0   

      VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0                    7.0                 2.0         20.0         4.0   
1                   10.0                 5.0         16.0         6.0   
2                   10.0                10.0          8.0         3.0   
3                   10.0                 7.0         12.0         4.0   
4                   10.0                 7.0         10.0         2.0   
...                  ...                 ...          ...         ...   
1314                10.0                10.0         12.0         5.0   
1315                10.0                10.0         14.0         5.0   
1316                10.0                10.0         12.0         4.0   
1317                10.0                10.0         13.0         4.0   
1318                10.0                10.0         12.0         4.0   

      WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0            31.0     0    0     1             1  
1            25.0     0    0     0             0  
2            12.0     0    0     0             0  
3            20.0     0    0     0             0  
4            16.0     0    0     0             0  
...           ...   ...  ...   ...           ...  
1314         21.0     0    0     0             0  
1315         20.0     0    0     0             0  
1316         17.0     0    0     0             0  
1317         20.0     0    0     0             0  
1318         20.0     0    0     0 

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [126]:
from sklearn.model_selection import train_test_split


Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [127]:
#1. 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state=0)

In [128]:
# 2. 

X_train

Date  TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
1261  736333         91        68           71.0          69.0          66.0   
49    735241         65        37           42.0          36.0          29.0   
674   735624         79        58           57.0          54.0          50.0   
1191  736355         86        63           66.0          61.0          51.0   
549   735621         92        75           76.0          73.0          71.0   
...      ...        ...       ...            ...           ...           ...   
763   735986         57        31           32.0          30.0          27.0   
835   735967         83        53           53.0          46.0          38.0   
1216  736349         88        70           68.0          67.0          65.0   
559   735601         91        76           74.0          72.0          69.0   
684   735603         85        72           74.0          72.0          69.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
1261                100.0                76.0                52.0   
49                   85.0                63.0                40.0   
674                  78.0                62.0                45.0   
1191                 93.0                67.0                40.0   
549                  94.0                73.0                52.0   
...                   ...                 ...                 ...   
763                  85.0                61.0                36.0   
835                  66.0                48.0                30.0   
1216                 90.0                73.0                55.0   
559                  88.0                69.0                49.0   
684                 100.0                83.0                65.0   

      SeaLevelPressureHighInches  ...  VisibilityHighMiles  \
1261                       29.90  ...                 10.0   
49                         30.26  ...                 10.0   
674                        30.05  ...                 10.0   
1191                       29.91  ...                 10.0   
549                        30.18  ...                 10.0   
...                          ...  ...                  ...   
763                        30.37  ...                 10.0   
835                        30.21  ...                 10.0   
1216                       30.07  ...                 10.0   
559                        30.03  ...                 10.0   
684                        30.02  ...                 10.0   

      VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
1261                 9.0                 2.0         12.0         1.0   
49                   7.0                 3.0         14.0         5.0   
674                 10.0                10.0         15.0         6.0   
1191                 8.0                 2.0         18.0         8.0   
549                 10.0                10.0         13.0         5.0   
...                  ...                 ...          ...         ...   
763                 10.0                10.0          8.0         3.0   
835                 10.0                10.0         12.0         4.0   
1216                10.0                10.0         12.0         6.0   
559                  9.0                 7.0         12.0         6.0   
684                  9.0                 2.0         12.0         6.0   

      WindGustMPH  Snow  Fog  Rain  Thunderstorm  
1261         24.0     0    0     1             1  
49           21.0     0    0     0             0  
674          23.0     0    0     0             0  
1191         27.0     0    0     0             0  
549          20.0     0    0     0             1  
...           ...   ...  ...   ...           ...  
763          12.0     0    0     0             0  
835          18.0     0    0     0             0  
1216         20.0     0    0     0             0  
559          19.0     0    0     0             0  
684          19.0     0    0     1 

In [131]:
len(X_train)

1055

In [138]:
X_train_percentage = (len(X_train)/len(austin_transformed_fixed))*100
X_train_percentage

79.98483699772555

In [129]:
# 3. 

X_test

Date  TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
658   735608         90        67           62.0          59.0          58.0   
406   735629         62        49           61.0          51.0          37.0   
1202  736335         79        49           47.0          43.0          40.0   
202   735244         95        71           72.0          67.0          56.0   
1187  736351         83        68           64.0          62.0          58.0   
...      ...        ...       ...            ...           ...           ...   
608   735619         92        72           74.0          71.0          67.0   
194   735236         96        72           71.0          69.0          61.0   
240   735251        102        78           75.0          71.0          64.0   
1154  736346         77        50           60.0          50.0          41.0   
1275  736347        100        78           74.0          71.0          67.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
658                  78.0                56.0                34.0   
406                 100.0                80.0                59.0   
1202                 77.0                52.0                27.0   
202                  90.0                60.0                30.0   
1187                 87.0                67.0                47.0   
...                   ...                 ...                 ...   
608                  93.0                73.0                52.0   
194                  93.0                62.0                30.0   
240                  85.0                59.0                33.0   
1154                 93.0                67.0                41.0   
1275                 85.0                60.0                35.0   

      SeaLevelPressureHighInches  ...  VisibilityHighMiles  \
658                        30.21  ...                 10.0   
406                        30.26  ...                 10.0   
1202                       30.29  ...                 10.0   
202                        30.15  ...                 10.0   
1187                       30.16  ...                 10.0   
...                          ...  ...                  ...   
608                        30.03  ...                 10.0   
194                        30.13  ...                 10.0   
240                        29.99  ...                 10.0   
1154                       29.98  ...                 10.0   
1275                       29.82  ...                 10.0   

      VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
658                 10.0                10.0         10.0         3.0   
406                  5.0                 1.0         15.0         4.0   
1202                10.0                10.0          9.0         3.0   
202                 10.0                10.0         13.0         5.0   
1187                10.0                10.0         14.0         7.0   
...                  ...                 ...          ...         ...   
608                  9.0                 4.0         14.0         5.0   
194                 10.0                 6.0         20.0         3.0   
240                 10.0                 9.0         15.0         7.0   
1154                10.0                10.0         13.0         6.0   
1275                10.0                10.0         13.0         7.0   

      WindGustMPH  Snow  Fog  Rain  Thunderstorm  
658          16.0     0    0     0             0  
406          21.0     0    0     1             0  
1202         15.0     0    0     0             0  
202          22.0     0    0     0             0  
1187         23.0     0    0     0             0  
...           ...   ...  ...   ...           ...  
608          23.0     0    0     1             0  
194          29.0     0    0     1             1  
240          24.0     0    0     0             0  
1154         22.0     0    0     0             0  
1275         23.0     0    0     0 

In [135]:
len(X_test)

264

In [139]:
X_test_percentage = (len(X_test)/len(austin_transformed_fixed))*100
X_test_percentage

20.01516300227445

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [ ]:
# Your code here:
